### PR TITLE
Update staff semester view

### PR DESCRIPTION
**Pull request by janno42**
This changes the layout of the staff semester view. All courses are now shown in a single datatable. This enables custom sorting. Also, additional information was added to the evaluation period column and there is a new option for approving a course without saving changes.
Approving courses is now only possible from the course page.

staffsemester

### CODE REVIEW

**Comment by janno42** [comment]
1. that was a bug i noticed but didn't know how to fix. actually it's fine now with a default ordering :)
2. default order is now by course name.
3. that's intentional: courses where staff users have to do something come first.
4. that was my initial plan. i thought about a progress bar representing the states and making that sortable. when it came to the comment review, i noticed that it's not a linear process and actually as a staff user i'm not that much interested in the internal state but more in the open tasks for a course. comment reviews start already when the course is still in evaluation and i want to be able to sort by that.
5. gray icons are now "Evaluation is not finished". Yellow and red are basically the same (technically the results could be published) but red indicates, that maybe someone just forgot to publish the results of a single result or that the grades are already available but the results were still not published.
6. will discuss (see 10). i had text there in the beginning but thought it might be unneeded clutter. after all this page does not need to be self-explanatory on first view, but should optimize on efficient workflow for everyday staff use.
7. an additional note or confirmation would be nice, i agree. the option is nice because it allows to approve without being the last user who changed the course in case of later occurring questions to the last real editor. additionally, the review sometimes includes e.g. checking the calendar by clicking on the evaluation period dates without intention to change anything and makes it easy to just click around without the fear to actually change any data.
8. green means "done", yellow "waiting for staff action", gray "currently not of interest", red "nothing happened yet/something sees wrong" - always focused on staff tasks. actually this was wrong for the evaluation icon, so i changed it as follows: gray pause symbol until evaluation starts, then a gray play-triangle (both analogous to the evaluation period counter symbol), green stop-square once evaluation is done.
9. see 11
10. in progress. there might be some smaller changes after discussing the layout with the one other person currently actively using this page.
11. i'm not really sure about the naming. what would you suggest? "new" doesn't really explain anything and a tooltip "this course is new" sounded like advertisement ;)
12. yes, done.

**Comment by janno42** [comment]
6: text is not needed.
10: done. a few things have changed after the feedback: the semester action buttons were moved (so the action card could completely be removed) and a new filtering option was added to the table. this makes the datatable creation a little bit more complex but is probably something similar to what we have in mind for the results page.

staff

**Comment by janno42** [comment]
so i think left for discussion are (at least)
- 7 (confirmation when approving without saving)
- 9, 11 (naming of states)

**Comment by karyon** [comment]
* 7. my first thought would be that users shouldn't change anything if they don't intend to change anything. but okay, that might happen accidentally. in that case they can press ctrl-z for the non-js-fields, and they cant press ctrl-f5 to reload the page and discard the entered form data, but okay, that's non-obvious. if you really think that functionality is necessary, i would be fine with making that button a more dangerous color and label it "discard changes and approve".

* 9, 11: i'm fine with giving the states more verbose names in the UI than they have in the code, e.g. `awaiting editor approval` is probably not gonna be what the state is called in the code. i'm also fine with calling the states slightly differently in different contexts, e.g. `approved by editor, awaiting staff review` might be too long for some locations. but i think the UI should be consistent, e.g. `reverted 1 course to new` is just plain inconsistent when `new` appears anywhere else. e.g. `not yet prepared` would be fine in that context as well wouldn't it? my approach in that regard would be to make the UI consistent first and then we can consider changing the state names in the code. 


[]()
 
13. right now i see only 10 courses with no filters selected, although the semester contains 21 courses (test data, active semester). it looks like the selection corresponds to the "evaluation did not start yet" filter.
13. when activating the filters, the browser scrollbar disappears sometimes, causing the webpage to shift. maybe we should [always show the scrollbar](https://stackoverflow.com/questions/4050076/how-to-always-show-the-vertical-scrollbar-in-a-browser) on pages with datatables?
13. the filter buttons have standard tooltips, not the styled ones.
13. the tooltips on the symbols on the buttons at the bottom seem to not work on firefox
13. why red for "not yet prepared"?
13. i thought courses that are not graded or courses with final grades get automatically published right away?
13. i was expecting to be able to select multiple filters, but if you think that doesn't help i'm fine with that.

**Comment by janno42** [comment]
7: actually the use case more important to me is the one mentioned first: if the staff user knows that no changes are made and just wants to approve the course (which is not a "change" of the data in my perception), then this option allows to do just that and keep the last real editor as is. so it's the button i would use at least as often as the other one (if editors would make no mistakes i would solely use this one) - so it would be nice not making it to "dangerously looking". i'm sure we'll find something nice.

9, 11: something like in the new commit?

13: oops. that was pagination applied to the table without showing page buttons. it was a mistake during the code shortening before my commit - fixed now.

14: how about always showing the scrollbar on all pages all the time (like in the current version)? then we would never have jumping pages on any page change or refresh.

15: fixed.

16: true. firefox does not show tooltips in elements inside buttons. i didn't find a nice solution so far. replacing the buttons with something else would require javascript for sending the form. placing the icons in front of the buttons looks bad. using just one tooltip for the button itself would result in longer texts for e.g. the button with three icons. any ideas?

17: could also be something different, i'm open to suggestions. i probably chose red to indicate that this is a state where "something needs to be done to get things on the track".

18: only if all comments have already been reviewed once the evaluation period ends. otherwise it has to be done manually (to allow publishing multiple courses simultaneously so that they can be included in the same notification email).

19: i tested this version as well. it would only be used in very few special use cases and for those the course list can just be sorted to have the important courses on top. i chose to not allow it because it would be a combination of `or` (in groups) and `and` (between groups) filtering that might be confusing, most combinations would have empty results and and additional button for "remove filters" would be needed.

File: evap/staff/templates/staff_semester_view_course.html [hunk 1]
```diff
@@ -16,114 +78,177 @@
...
-            </span>
+            {% if not course.can_publish_grades %}
+                <span data-toggle="tooltip" data-placement="left" title="{% trans 'Not enough voters to publish results' %}">
+                    {% include "progress_bar.html" with done=course.num_voters total=course.num_participants icon="user-o" warning=True %}
+                </span>
+            {% else %}
+                <span data-toggle="tooltip" data-placement="left" title="{% trans 'Number of voters' %}">
+                    {% include "progress_bar.html" with done=course.num_voters total=course.num_participants icon="user-o" %}
+                </span>
+            {% endif %}
```
Comment (karyon) [inline evap/staff/templates/staff_semester_view_course.html]: swap the two cases?

File: evap/staff/tests/test_views.py [hunk 2]
```diff
@@ -928,6 +912,65 @@ def test_remove_responsibility(self):
...
+        form["vote_start_datetime"] = new_vote_start_datetime
+        form["vote_end_date"] = new_vote_end_date
+        form["name_de"] = new_name_de
+        form.submit(name="operation", value="approve")
+
+        self.course = Course.objects.get(pk=self.course.pk)
+        self.assertEqual(self.course.name_de, old_name_de)
+        self.assertEqual(self.course.vote_start_datetime, old_vote_start_datetime)
+        self.assertEqual(self.course.vote_end_date, old_vote_end_date)
+        self.assertEqual(self.course.state, 'approved')
```
Comment (karyon) [inline evap/staff/tests/test_views.py]: shouldn't the se be `NotEqual`?

File: evap/static/css/evap.scss [hunk 3]
```diff
@@ -370,6 +377,7 @@ body {
     background-image: url("../images/triangles_gray.svg");
     background-size: cover;
     background-attachment: fixed;
+    overflow-y: scroll !important;
```
Comment (karyon) [inline evap/static/css/evap.scss]: please put that into a separate commit

File: evap/staff/tests/test_views.py [hunk 4]
```diff
@@ -912,66 +913,42 @@ def test_remove_responsibility(self):
...
-        form["name_de"] = new_name_de
+        # approve without changes
         form.submit(name="operation", value="approve")
 
         self.course = Course.objects.get(pk=self.course.pk)
-        self.assertEqual(self.course.name_de, old_name_de)
-        self.assertEqual(self.course.vote_start_datetime, old_vote_start_datetime)
-        self.assertEqual(self.course.vote_end_date, old_vote_end_date)
+        self.assertEqual(self.course.last_modified_user, old_last_modified_user)
         self.assertEqual(self.course.state, 'approved')
```
Comment (karyon) [inline evap/staff/tests/test_views.py]: can you assert the other values didn't change?

File: evap/staff/tests/test_views.py [hunk 5]
```diff
@@ -912,66 +913,42 @@ def test_remove_responsibility(self):
...
+        form["name_de"] = "Test name"
+        # approve after changes
+        form.submit(name="operation", value="approve")
 
         self.course = Course.objects.get(pk=self.course.pk)
-        self.assertEqual(self.course.name_de, new_name_de)
-        self.assertEqual(self.course.vote_start_datetime, datetime.datetime.strptime(new_vote_start_datetime, '%Y-%m-%d %H:%M'))
-        self.assertEqual(self.course.vote_end_date, datetime.datetime.strptime(new_vote_end_date, '%Y-%m-%d').date())
+        self.assertEqual(self.course.last_modified_user, test_user)
         self.assertEqual(self.course.state, 'approved')
```
Comment (karyon) [inline evap/staff/tests/test_views.py]: can you assert the other values did change?

**Comment by karyon** [comment]
can you try to squash some of the commits? as far as i can see, most commits apart from the 
`relocate action buttons and remove action card` and `add filter options to course table in staff semester view` have little value. `add option to approve course without saving` was reverted, but the change to that form would be nice to have in a separate commit actually.

File: evap/staff/templates/staff_semester_view_course.html [hunk 6]
```diff
@@ -16,114 +78,177 @@
...
-        <br>
-        {% if course.days_left_for_evaluation <= 0 %}
-            <span class="badge badge-danger">{% trans "ends today" %}</span>
-        {% elif course.days_left_for_evaluation == 1 %}
-            <span class="badge badge-danger">{% trans "ends tomorrow" %}</span>
-        {% elif course.days_left_for_evaluation < 5 %}
-            <span class="badge badge-secondary">{% blocktrans with days=course.days_left_for_evaluation %}{{ days }} days left{% endblocktrans %}</span>
+>
+    {% if course.is_single_result %}
+        {{ course.vote_start_datetime|date:"Y-m-d"}}
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view_course.html]: because single results just have an event date, no starting time and no end date.

File: evap/evaluation/templatetags/evaluation_filters.py [hunk 7]
```diff
@@ -10,6 +10,11 @@ def zip_lists(a, b):
     return zip(a, b)
 
 
+@register.filter(name='reverse_order')
+def reverse_order(value):
+    return 999999-abs(value)
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/evaluation/templatetags/evaluation_filters.py]: nice. it's now
```
if course.state in ['new', 'prepared', 'editor_approved', 'approved']:
        return course.days_until_evaluation
    elif course.state == "in_evaluation":
        return 100000 + course.days_left_for_evaluation
    else:
        return 200000 + course.days_left_for_evaluation
```
(higher numbers because in the test data we have dates in 2099, so the number of days can reach up to 30,000)

File: evap/staff/templates/staff_semester_view.html [hunk 8]
```diff
@@ -59,116 +60,110 @@ <h3>
...
-                                <button name="operation" value="{{state}}->published" type="submit" class="btn btn-secondary" {{ disable_if_archived }}>{% trans "Publish courses" %}</button>
-                            {% endifequal %}
-                            {% ifequal state "published" %}
-                                <button name="operation" value="{{state}}->reviewed" type="submit" class="btn btn-secondary" {{ disable_if_archived }}>{% trans "Unpublish" %}</button>
-                            {% endifequal %}
-                            {% endif %}
-                        </form>
+                {% if request.user.is_staff %}
+                    <div class="my-2">
+                        <a role="button" class="btn btn-sm btn-secondary {{ disable_if_archived }}"
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: check

File: evap/staff/templates/staff_semester_view.html [hunk 9]
```diff
@@ -274,5 +239,88 @@ <h3>
             var state = window.localStorage['collapse_overview'] === 'true';
             localStorage['collapse_overview'] = (!state).toString();
         }
+
+        $(window).bind("pageshow", function () {
+            $('#course_operation_form :checkbox').prop('checked', false); // prevent false autofills
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: check

File: evap/staff/templates/staff_semester_view.html [hunk 10]
```diff
@@ -274,5 +239,88 @@ <h3>
...
+        <script type="text/javascript" src="{% get_static_prefix %}js/plugins/jquery.dataTables.min.js"></script>
+        <script type="text/javascript" src="{% get_static_prefix %}js/plugins/dataTables.bootstrap4.min.js"></script>
+    {% endcompress %}
+
+    <script type="text/javascript">
+        $.fn.dataTable.ext.search.push(
+            function( settings, data, dataIndex ) {
+                if ($("[data-filter].active").length > 0){
+                    // check whether any icon row contains the data-filter value of the active button
+                    return data[1].concat(data[2], data[3], data[4]).includes($("[data-filter].active")[0].dataset.filter);
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: maybe you should do the js-parts in the future ;)

File: evap/staff/templates/staff_semester_view.html [hunk 11]
```diff
@@ -274,5 +239,88 @@ <h3>
...
+                }
+                return true;
+            }
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: check

File: evap/staff/templates/staff_semester_view.html [hunk 12]
```diff
@@ -274,5 +239,88 @@ <h3>
...
+                                    data-toggle="tooltip" data-placement="top" title="{% trans 'Results published' %}"></a>
+                            </div>
+                        </div>`
+                    );
+                    $('[data-filter]').click( function(e) {
+                        if($(e.target).hasClass("active")) {
+                            $("[data-filter].active").removeClass("active");
+                        }
+                        else {
+                            $("[data-filter].active").removeClass("active");
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: now
```
var make_active = !$(e.target).hasClass("active");
    $("[data-filter].active").removeClass("active");
    if (make_active) {
        $(e.target).addClass("active");
    }
```

File: evap/staff/templates/staff_semester_view_course.html [hunk 13]
```diff
@@ -2,12 +2,74 @@
...
+    {% else %}
+        <td class="icon text-center" data-order="1"><span class="fa fa-bar-chart icon-yellow" data-toggle="tooltip" data-placement="top" title="{% trans 'Results not yet published' %}"></span></td>
+    {% endif %}
+{% elif state == "published" %}
+    <td class="icon" data-order="3"><span class="fa fa-check-square-o icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Approved by staff' %}"></span></td>
+    <td class="icon" data-order="3"><span class="fa fa-stop icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Evaluated' %}"></span></td>
+    <td class="icon" data-order="3"><span class="fa fa-comment icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Text answers reviewed' %}"></span></td>
+    <td class="icon" data-order="3"><span class="fa fa-bar-chart icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Results published' %}"></span></td>
+{% endif %}
+<td style="width: {% if info_only %}51%{% else %}42%{% endif %}" data-order="{{ course.name }}">
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view_course.html]: info_only does not include buttons at the end of the line and the link for reviewing text answers is disabled, this is not the case for reviewers, so there are really three different views for the page.

File: evap/staff/templates/staff_semester_view_course.html [hunk 14]
```diff
@@ -2,12 +2,74 @@
...
+{% if not info_only %}
+    {% if not request.user.is_staff %}
+        <td></td>
+    {% elif state == "new" or state == "prepared" or state == "editor_approved" or state == "approved" or state == "reviewed" or state == "published" %}
+        <td>
+            <input type="checkbox" name="course" id="course{{course.id}}" value="{{course.id}}" {{disable_if_archived}} data-toggle="tooltip"
+                data-placement="top" title="{% trans 'Click to select for course operation' %}" />
+        </td>
+    {% else %}
+        <td></td>
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view_course.html]: now
```
{% if info_only or semester_is_archived or not request.user.is_staff %}
    <td></td>
{% elif state == "new" or state == "prepared" or state == "editor_approved" or state == "approved" or state == "reviewed" or state == "published" %}
    <td>
        <input type="checkbox" name="course" id="course{{course.id}}" value="{{course.id}}" data-toggle="tooltip"
            data-placement="top" title="{% trans 'Click to select for course operation' %}" />
    </td>
{% else %}
    <td></td>
{% endif %}
```

File: evap/staff/templates/staff_semester_view_course.html [hunk 15]
```diff
@@ -2,12 +2,74 @@
...
+        <td class="icon text-center" data-order="0"><span class="fa fa-bar-chart icon-red" data-toggle="tooltip" data-placement="top" title="{% trans 'Results not yet published' %}"></span></td>
+    {% else %}
+        <td class="icon text-center" data-order="1"><span class="fa fa-bar-chart icon-yellow" data-toggle="tooltip" data-placement="top" title="{% trans 'Results not yet published' %}"></span></td>
+    {% endif %}
+{% elif state == "published" %}
+    <td class="icon" data-order="3"><span class="fa fa-check-square-o icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Approved by staff' %}"></span></td>
+    <td class="icon" data-order="3"><span class="fa fa-stop icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Evaluated' %}"></span></td>
+    <td class="icon" data-order="3"><span class="fa fa-comment icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Text answers reviewed' %}"></span></td>
+    <td class="icon" data-order="3"><span class="fa fa-bar-chart icon-green" data-toggle="tooltip" data-placement="top" title="{% trans 'Results published' %}"></span></td>
+{% endif %}
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view_course.html]: check

File: evap/staff/templates/staff_semester_view_course.html (hunk 1 above)
Comment (janno42) [reply evap/staff/templates/staff_semester_view_course.html]: check

File: evap/static/css/evap.scss (hunk 3 above)
Comment (janno42) [reply evap/static/css/evap.scss]: will do

File: evap/staff/tests/test_views.py (hunk 4 above)
Comment (janno42) [reply evap/staff/tests/test_views.py]: is it okay to only test some values (e.g. name_de, vote_start_datetime, vote_end_date)?

**Comment by janno42** [comment]
i addressed your comments in a further commit and will squash and reorder the commits if everything else is okay.

File: evap/staff/templates/staff_semester_view_course.html (hunk 14 above)
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view_course.html]: nice!

**Comment by karyon** [comment]
as far as i remember, stuff was only disabled instead of removed if a semester is archived. you removed some of the buttons, why not go all the way and remove everything actionable and remove the `disable_if_archived`?

**Comment by janno42** [comment]
instead of `disable_if_archived` the respective buttons have now been removed.

**Comment by karyon** [comment]
can you do the squashing? i would like to review again and having proper commits would be a great help.

**Comment by janno42** [comment]
done

File: evap/staff/views.py [hunk 16]
```diff
@@ -157,55 +142,71 @@ def semester_course_operation(request, semester_id):
...
     email_template = None
     if courses:
-        if target_state == 'prepared':
-            email_template = EmailTemplate.objects.get(name=EmailTemplate.EDITOR_REVIEW_NOTICE)
-
-        elif target_state == 'approved':
-            # remove courses without questionnaires on general contribution, warn about courses with missing questionnaires
-            courses_with_enough_questionnaires = [course for course in courses if course.general_contribution_has_questionnaires]
-            courses_with_missing_questionnaires = [course for course in courses_with_enough_questionnaires if not course.all_contributions_have_questionnaires]
+        if target_state == 'new':
```
Comment (karyon) [inline evap/staff/views.py]: previously there wasn't any code for this? why?

File: evap/staff/views.py [hunk 17]
```diff
@@ -157,55 +142,71 @@ def semester_course_operation(request, semester_id):
...
-                messages.warning(request, ungettext("%(courses)d course can not be approved, because it's evaluation end date lies in the past. It was removed from the selection.",
-                    "%(courses)d courses can not be approved, because their evaluation end dates lie in the past. They were removed from the selection.",
+                courses = courses_ready_for_evaluation
+                messages.warning(request, ungettext("The evaluation for %(courses)d course can not be started, because it was not approved, was already evaluated or its evaluation end date lies in the past. It was removed from the selection.",
+                    "The evaluation for %(courses)d courses can not be started, because they were not approved, were already evaluated or their evaluation end dates lie in the past. They were removed from the selection.",
                     difference) % {'courses': difference})
             email_template = EmailTemplate.objects.get(name=EmailTemplate.EVALUATION_STARTED)
+            confirmation_message = _("Do you want to immediately start the evaluation for the following courses?")
+
+        elif target_state == 'reviewed':
```
Comment (karyon) [inline evap/staff/views.py]: same here?

File: evap/staff/views.py [hunk 18]
```diff
@@ -157,55 +142,71 @@ def semester_course_operation(request, semester_id):
             helper_semester_course_operation_revert(request, courses)
         elif target_state == 'prepared':
             helper_semester_course_operation_prepare(request, courses, template)
-        elif target_state == 'approved':
-            helper_semester_course_operation_approve(request, courses)
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/views.py]: because the operation does not exist anymore - courses can only be approved from the course form.

File: evap/staff/views.py (hunk 16 above)
Comment (janno42) [reply evap/staff/views.py]: we now have buttons for all course operations on one page. not all operations are valid for courses in all states. we needed less checks before, because the available operations were per state.

File: evap/evaluation/models.py [hunk 19]
```diff
@@ -459,9 +463,6 @@ def warnings(self):
                 result.append(_("Course has no questionnaires"))
             else:
                 result.append(_("Not all contributors have questionnaires"))
-
-        if self.state in ['in_evaluation', 'evaluated', 'reviewed', 'published'] and not self.can_publish_grades:
-            result.append(_("Not enough participants to publish results"))
         return result
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/evaluation/models.py]: done

File: evap/staff/templates/staff_semester_view.html [hunk 20]
```diff
@@ -110,62 +110,75 @@ <h3>
...
+                            <th>{% trans 'Course' %}</th>
+                            <th>{% trans 'Evaluation period' %}</th>
+                            <th data-orderable="false"></th>
+                            <th data-orderable="false"></th>
+                        </tr>
+                    </thead>
+                    <tbody>
+                    {% for course in courses %}
+                        <tr id="course-row-{{ course.id }}">
+                            {% include 'staff_semester_view_course.html' with semester=semester course=course semester_is_archived=semester.is_archived info_only=False %}
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: check

File: evap/staff/templates/staff_semester_view.html [hunk 21]
```diff
@@ -271,5 +254,92 @@ <h3>
...
+                                    data-toggle="tooltip" data-placement="top" title="{% trans 'Results published' %}"></a>
+                            </div>
+                        </div>`
+                    );
+                    $('[data-filter]').click( function(e) {
+                        var make_active = !$(e.target).hasClass("active");
+                        $("[data-filter].active").removeClass("active");
+                        if (make_active) {
+                            $(e.target).addClass("active");
+                        }
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/staff/templates/staff_semester_view.html]: changed to toggle and not
